### DISCUSSION TITLE
How to prepend `\<C-U>` to `vim.cmd` API?

### DISCUSSION THREAD

**Discussion by linrongbin16** [neovim/neovim#24055] [Q&A] [answered]
Hi, I am writing a key map util function:

```lua
local function exec(op)
    if type(op) == "string" then
        -- vim command
        vim.cmd(op)
    else
        -- lua function
        op()
    end
end

local function map(mode, lhs, rhs, opts)
    opts = vim.tbl_deep_extend("force", {silent=true, noremap=true}, opts or {})
    vim.keymap.set(mode, lhs, rhs, opts)
end

-- here define some key mappings

map({'n', 'x'}, '<leader>f', exec('echo hello'))
```

But now, I have a specific requirement, that I want to get the block-wise visual selection. based on this: https://stackoverflow.com/a/61486601/4438921, I found that I have to use vimscript to define key mapping:

```vim
xnoremap <leader>f :\<C-U> call GetVisualSelection(visualmode())<cr>
```

How could I do it in lua?

**Comment by oncomouse** [comment]
I think `vim.keymap.set({'x'}, ":<c-u> call GetVisualSelection(visualmode())<cr>")` should work.  You don't need to wrap `rhs` in that exec function inside of your `map()` function (which you also don't need). `vim.keymap.set` already correctly handles strings and functions passed to `rhs`.

**Reply by linrongbin16** [reply]
emmmm, but how should I map it in lua way? If now I have a vim function called `fzf#vim#grep`, or a vim command called `FzfGrep`, how should I define it in lua API `vim.keymap.set` rhs?

**Reply by oncomouse** [reply]
`vim.keymap.set({'x'}, '<leader>f', function() vim.fn.GetVisualSelection(vim.fn.visualmode()) end)` might work if you really want to do it with a function.

**Reply by oncomouse** [reply]
With the command, you can call `vim.cmd('FzfGrep')`. With the other function, you can run `vim.fn['fzf#vim#grep']()`

**Reply by linrongbin16** [reply]
In that case, it seems I will miss the `\<C-U>`, thus I cannot get the correct block-wise visual selection?

**Reply by linrongbin16** [reply]
mayby I can try adding a `normal!` before calling the function: `vim.cmd('normal! call getVisualSelection(visualmode())')`.


because based on another answer: https://stackoverflow.com/a/47051271/4438921,  it says
```
`'<` and `'>` don't get updated when the user is still in visual mode.
```

That is the reason, that I cannot get the correct visual selection in `getVisualSelection` function.

But if now I try to add `normal!`, maybe it will first exit visual mode, then I can get the correct (e.g. the previous visual mode) visual selection

**Reply by TheLeoP** [reply, marked as answer]
> In that case, it seems I will miss the \<C-U>, thus I cannot get the correct block-wise visual selection?

In this context, those keys only delete `'<,'>` because it gets automatically inserted in the command line when you press `:` on visual mode. If you check `:h c_ctrl-u` you will see that this is what `<c-u>` does in command mode.

> mayby I can try adding a normal! before calling the function: vim.cmd('normal! call getVisualSelection(visualmode())').
> because based on another answer: https://stackoverflow.com/a/47051271/4438921, it says
> `'<` and `'>` don't get updated when the user is still in visual mode.
> That is the reason, that I cannot get the correct visual selection in getVisualSelection function.
> But if now I try to add normal!, maybe it will first exit visual mode, then I can get the correct (e.g. the previous visual mode) visual selection

Yes, those marks are not set until you leave visual mode. But, `\<c-u>` has nothing to do with it, the thing that makes you leave visual mode in the vimscript mapping is `:` (because you enter command mode). So, if you wan't to force neovim to leave visual mode so the marks are set, you can use `vim.cmd.normal('')` (where `` is what neoivm sees when you press `<esc>`). The full function would be:

```lua
vim.keymap.set({ "x" }, "<leader>f", function()
  vim.cmd.normal ""
  vim.fn.GetVisualSelection(vim.fn.visualmode())
end)
```

**Reply by linrongbin16** [reply]
Oh yes, what you said is really that I want!
BTW, can I use `\<ESC>` instead of `�`? I don't how to write it in my config.

**Reply by linrongbin16** [reply]
I tried `<ESC>` in this commit: https://github.com/linrongbin16/lin.nvim/pull/127/commits/f63d5b4c86c5084ec66e43c0f822db4147e2b6fa

But there's an error message:
<img width="211" alt="image" src="https://github.com/neovim/neovim/assets/6496887/d94425d4-7833-441f-98fa-948af39141a8">

Looks like there's some escape char issue....

**Reply by linrongbin16** [reply]
Finally I fixed it in commit: https://github.com/linrongbin16/lin.nvim/pull/128/commits/86f929eba219a3655f3db1edf873a0c059b66924
